### DISCUSSION TITLE
How to send data from MainActivity to React Native side

### DISCUSSION THREAD

**Discussion by BraveEvidence** [reactwg/react-native-new-architecture#176] [Q&A]
I have enabled the new architecture for android using React Native v0.74. I need to pass some data from MainActivity to React Native.  Previously I used to use something like below code

```
val context = this.reactNativeHost.reactInstanceManager.currentReactContext
context!!.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)?.emit("result", "data:image/png;base64," + Base64.encodeToString(byteArray, Base64.DEFAULT))
```

But I get a crash saying `context` is null, this way used to work with React Native 0.73 for new and old arch, what is the recommended approach now

**Comment by cortinico** [comment]
> what is the recommended approach now

You would have to use the EventDispatcher.
Example on how to use it is here: https://github.com/react-native-menu/menu/pull/749

**Reply by Bardiamist** [reply]
@cortinico Hey, example is difficult, do we have simple example for MainActivity? Would be good to have it in the documentation.

**Comment by BraveEvidence** [comment]
@cortinico  Thanks for the reply. The example you mentioned is of Fabric Components. Is the approach same for sending string data from MainActivity if I don't have a Fabric Component. 

Also, I did try the approach yo recommended 


```
class SendDataEvent(
    surfaceId: Int,
    viewTag: Int,
    private val event: String?,
    private val target: Int
) : Event<SendDataEvent>(surfaceId, viewTag) {

    override fun getEventName(): String = "onPressAction"
    override fun getEventData(): WritableMap = Arguments.createMap().apply {
        if (event != null) {
            putString("event", event)
        }
        putString("target", target.toString())
    }
}


 val id = 1
                val dispatcher =
                    UIManagerHelper.getEventDispatcherForReactTag(context,id)
                dispatcher?.dispatchEvent(SendDataEvent(UIManagerHelper.getSurfaceId(this), id, "data:image/png;base64," + Base64.encodeToString(byteArray, Base64.DEFAULT), id))
```

Yo can please guide on how do I listen to this in my react component

**Reply by lovegaoshi** [reply]
I dont think fabric or not matters..?
migration from RCTEventEmitter to dispatchEvent:
https://github.com/lovegaoshi/react-native-svga-player/pull/3/files#diff-9385651304063e779a30875dff3f371ee25644eb3430e8658e900acf11c787e8R28

event listener in js:
https://github.com/lovegaoshi/react-native-svga-player/blob/588c9a2819cb2fe14dd7204addaeadd4bf049333/index.js#L56

not that I really know if any of this is right, but my lib seems working still

**Reply by BraveEvidence** [reply]
@lovegaoshi Thanks for the reply, can you tell how to get the react context in MainActivity. I am using below code 

` val context = this.reactNativeHost.reactInstanceManager.currentReactContext ` 

but the context always is null. Any other way to get the context in MainActivity

**Reply by lovegaoshi** [reply]
i dont think this has to do with the new arch, bjut if u wanna try by all means please do
https://github.com/lovegaoshi/azusa-player-mobile/blob/3c8d25e4b3a19d541fe7be45b10b2fa1e38b349d/android/app/src/main/java/com/noxplay/noxplayer/MainActivity.kt#L87

**Reply by BraveEvidence** [reply]
@lovegaoshi  The code is the same which I am using , I get the context null.

**Comment by fgrs** [comment]
I'm facing same issue, has anybody figured out? What's the proper (new architecture) way?

**Reply by Mxrk** [reply]
I don't know if that's the preferred way, but migrating from author's way to
```val reactContext = this.reactActivityDelegate.reactHost.currentReactContext``` worked for me.
